### ISSUE TITLE
csgraph.min_weight_full_bipartite_matching restricted to full matches?

### ISSUE THREAD

**Issue by yalwan-sage** [scipy/scipy#14041]
I have some data which doesn't permit full matching. If I convert it to dense and solve using `scipy.optimize.linear_sum_assignment` I can get a (non-full) solution. 

Is it possible to remove this limitation of the csgraph solver? Why is that limitation there currently? I'm trying to expand to problem sizes which wouldn't be solvable using dense representations.

**Comment by pmla** [comment]
@fuglede perhaps you could comment?

**Comment by fuglede** [comment]
Without assuming fullness, you will generally not get anything useful: If all weights are positive (which is probably the common case), then the minimum weight bipartite matching is the empty one.

In particular, this is no different from what is happening in `linear_sum_assignment`; indeed, you can use either as a replacement of the other, and if there is an example for which the two provide different results, then that's a bug.

You mention that you "convert it to dense"; I wonder if you may have accidentally mapped missing edges to zero-cost edges in that process, and that's why you're seeing a difference? If not, would you be able share the example you have in mind?

**Comment by yaxxie** [comment]
> If all weights are positive (which is probably the common case), then the minimum weight bipartite matching is the empty one.

Understood -- however we can transform the edge weights and pose it as a maximisation.

> I wonder if you may have accidentally mapped missing edges to zero-cost edges

I can't give a definite answer right now, but I have taken a lot of care to remove all zero cost edges in the problem I am working on.

If I may, I'd like to give the following (carefully constructed, to illustrate the point) example:

```python
import scipy.sparse.csgraph
import scipy.optimize

d = [0.1, 0.2, 0.3, 0.4, 0.4, 0.3, 0.2, 0.1]
r = [0, 1, 2, 3, 4, 4, 4, 4]
c = [0, 0, 0, 0, 1, 2, 3, 4]
G = scipy.sparse.coo_matrix((d, (r, c)))

# number of connections we can make
print(sum(scipy.sparse.csgraph.maximum_bipartite_matching(G) > -1))

# illustrate the refusal to solve
try:
    scipy.sparse.csgraph.min_weight_full_bipartite_matching(G, maximize=True)
except ValueError:
    print("Sparse solver won't give a solution")

ridx, cidx = scipy.optimize.linear_sum_assignment(G.todense(), maximize=True)

# show the edge weights
G.todense()[ridx, cidx]

# show we have two candidate edges
(G.todense()[ridx, cidx] > 0).sum()
```

I apologise I didn't mention I was working with maximisation, but I'm sure you also appreciate that there isn't much separating the two conceptually (they're equivalent under some transformations).

**Comment by fuglede** [comment]
In your example, `G.dense()` becomes the complete graph, in which most edges have weight zero. In particular, in the output of `linear_sum_assignment(G.todense(), maximize=True)`, you will find that it has picked several non-existent edges.

If, instead, you use `-np.inf` (negative since we're maximizing), you'll find that it hits the same infeasibility check as its `csgraph` partner:

```python
In [11]: dense = np.full(G.shape, -np.inf)
    ...: dense[G.row, G.col] = G.data

In [12]: scipy.optimize.linear_sum_assignment(dense, maximize=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-8080a5287755> in <module>
----> 1 scipy.optimize.linear_sum_assignment(dense, maximize=True)

c:\users\sorfj\miniconda3\lib\site-packages\scipy\optimize\_lsap.py in linear_sum_assignment(cost_matrix, maximize)
    103         return (b[indices], a[indices])
    104     else:
--> 105         return _lsap_module.calculate_assignment(cost_matrix)

ValueError: cost matrix is infeasible
```

**Comment by yaxxie** [comment]
Ignoring all the problems with the response here, I think the answer to the question as posed is "no", is that correct?

**Comment by pmla** [comment]
@yaxxie yes, that is the short answer :)

`linear_sum_assignment` does not support non-full matches either - it raises an exception if a full solution cannot be found.

**Comment by yaxxie** [comment]
Thanks @fuglede for the time and responses.

**Comment by fuglede** [comment]
But just to make sure, what you were looking for is the maximum weight bipartite matching with the constraint that its cardinality should be the cardinality of a maximum bipartite matching? Because certainly that's not as trivial as I had in mind when I tried to describe what happens when you drop the fullness constraint altogether.

A more general problem is what is referred to in e.g. [Ramshaw, Tarjan -- *On Minimum-Cost Assignments in Unbalanced Bipartite Graphs*](https://www.hpl.hp.com/techreports/2012/HPL-2012-40R1.pdf) as the "imperfect assignment problem", which is about finding the maximum weight bipartite matching among all matchings of a given target size *t*; the special case in which *t* is the cardinality of a maximum bipartite matching falls into the case described in Section 1.3 of that paper, and can be solved through the doubling procedure described there as well. That would be straightforward enough to build on top of either of the existing two methods (although ine thing I don't immediately get is why they wouldn't need a cost of something like -4rC on the large-to-large linking edges), so that's absolutely something worth considering.

**Comment by yalwan-sage** [comment]
> But just to make sure, what you were looking for is the maximum weight bipartite matching among all bipartite matchings whose cardinality is the cardinality of a maximum bipartite matching?

Probably closer actually to the relaxation where we want a target size *t* (would that be exactly the target size or between *t* and the maximum cardinality?).

As the dicussion had evolved I realised that this is potentially a more research grade issue than I had anticipated. Thank you for sharing that report.

What form exactly of the more general problem do you think might be possible to build on top of the existing methods?

**Comment by yalwan-sage** [comment]
The other thing worth noting is that the example I shared was a graph with two connected components -- forming the problem properly would result in an infeasible solution, but solving for each connected component would work as expected, I think.

**Comment by fuglede** [comment]
> Probably closer actually to the relaxation where we want a target size t (would that be exactly the target size or between t and the maximum cardinality?).

It would be the exact size. (With the only notational caveat being that they allow *t* to exceed the largest cardinality, in which case they refer to the largest cardinality.)

> What form exactly of the more general problem do you think might be possible to build on top of the existing methods?

Just going by what's in the paper, only the maximum cardinality case: Let *r* be the cardinality of the smaller partition in a bipartite graph *G*, let *C* denote the largest absolute value of the weights, and let *ν*(*G*) be the cardinality of a maximum bipartite matching. Then the prescription in Section 1.3 is simplest when *t* = *ν*(*G*). Here, the problem amounts to replacing the biadjacency matrix `A` (with *r* rows) with something like the block matrix `[[A, 4rC Id], [0, A.T]]` which does have a perfect matching.

They do assume that the starting point is just a "good" algorithm for dealing with the case in which both partitions have the same cardinality, and since none of the two SciPy implementations are restricted to that case, perhaps it's possible to do even better.

**Comment by fuglede** [comment]
And yes, you're right, in your case you would also get the same result by considering each connected component. And in general, if you can assume that each connected component does have a full matching, then that's a simpler problem.

**Comment by yalwan-sage** [comment]
Do you think there's an appetite to have the algorithm described above or a variant as a non-full matcher in scipy csgraph/optimize? A user can in fact construct the structure themselves but it would be a useful convenience.

**Comment by fuglede** [comment]
Just for the maximum matching case, you mean? I imagine it would help that it's only a few lines of code (a few more for the sparse case) so it wouldn't become much of a maintenance burden. Hardest part would be coming up with a name that's better than `minimum_weight_maximum_bipartite_matching`. :)

**Comment by pmla** [comment]
I have reopened this issue. It seems I closed it prematurely - apologies :)

**Comment by yalwan-sage** [comment]
> Just for the maximum matching case, you mean?

Yes indeed, that probably fills most use cases above and beyond perfect matching.

> Hardest part would be coming up with a name that's better than minimum_weight_maximum_bipartite_matching

😅

**Comment by yalwan-sage** [comment]
@fuglede I tried to hack this atop the csgraph implementation using eps as the zeros, and it worked ok for small test cases I wrote up, but when I tried to run it with my real data it ended up falling into infinite loop. I presume this is some of the extra complexity with modifying the underlying solver that is required to support this? Is any work underway for this?

**Comment by fuglede** [comment]
@yalwan-sage Sounds good that this is still going!

* Using eps to avoid zeros being eliminated during conversion between types, you mean? The algorithm does play well with explicit zeros, so with a bit of care, it should be possible to avoid epsilons.
* Infinite loop sounds spooky; I'd expect the solver to always terminate, even if in an error if something went wrong. Would you be able to provide an example reproducing that behavior? It might prove to be a useful unit test. And useful to compare behavior with `linear_sum_assignment´, if the example isn't so large that it's impractical.

**Comment by yalwan-sage** [comment]
>The algorithm does play well with explicit zeros, so with a bit of care, it should be possible to avoid epsilons.

I added the epsilons because of the eliminate zeros call in the solver -- I wasn't fancying recompiling scipy with this line removed.

I'll get back to you on providing a repro

**Comment by yalwan-sage** [comment]
>  Zero weights can be handled by adding a constant to all weights, so that the resulting matrix contains no zeros.

I'll try again following this guidance.

**Comment by fuglede** [comment]
Ah, right, my bad regarding the explicit zeros; I was certain those were explicitly supported, but that's clearly not the case. Regardless of how you approach it, a reproducing example with epsilons could also be useful.

One thing to bear in mind is that in using what's efficiently rational numbers with very large denominators (as an epsilon would be), than can have an impact on augmenting path-style algorithms like both the one in `sparse.csgraph` and `optimize`; for example, Edmonds--Karp with rational weights has unbounded complexity. If that's what you're running into, then adding a "nice" constant (depending on your problem, something like 1), might be enough.

**Comment by yalwan-sage** [comment]
Sorry for little in the way of updates

I have managed to get the scheme to solve when I quantise my weights (before augmenting the graph) -- it seems like it may in fact be related to the unbounded complexity with rational weights issue. I will try to see if I can distill this down to a small example.

**Comment by yalwan-sage** [comment]
I haven't been able to find a small-ish reproducing example, but I'd like to share the implementation I've used for the augmenting form with dummy links, in case you find it useful and would want to incorporate a variant of it into scipy itself. The quantisation isn't incorporated because that's an issue of our specific problem rather than the general implementation.

```python
def minimum_weight_nonfull_matching(bigraph, maximise):

    rows, cols = bigraph.shape

    maxval = numpy.max(numpy.abs(bigraph))
    sentinel_value = 4 * maxval * rows  # TODO: find out if its unconditionally rows or min/max of the two sets
    if maximise:
        sentinel_value = -sentinel_value

    # DO NOT CALL `eliminate_zeros()` on this matrix
    upper_right_indices = numpy.arange(rows)
    upper_right_diag = scipy.sparse.coo_matrix((numpy.zeros(rows), (upper_right_indices, upper_right_indices)))

    lower_left_indices = numpy.arange(cols)
    lower_left_diag = scipy.sparse.coo_matrix(
        (numpy.zeros(cols) + sentinel_value, (lower_left_indices, lower_left_indices))
    )

    # DO NOT CALL `eliminate_zeros()` on this matrix
    new_bigraph = scipy.sparse.bmat(
        [[bigraph, upper_right_diag], [lower_left_diag, bigraph.T]],
        format="csr",  # might as well put it to CSR here, save a conversion
    )

    # this step eliminates zeros from the formulation for use with the solver
    new_bigraph.data = new_bigraph.data - numpy.min(new_bigraph.data) + 1

    # Make sure that we have the right number of elements present in fact
    assert new_bigraph.nnz == (bigraph.nnz * 2 + rows + cols)

    # solve the augmented formulation
    row_idx, col_idx = scipy.sparse.csgraph.min_weight_full_bipartite_matching(new_bigraph, maximize=maximise)

    # now pull out the subset of indices (cause dummy links are unneeded)
    # The summary here is that -1's are unassigned nodes and
    # anything from the row connected to bigger than an original colnode is a dummy link
    # and anything from the col connected to bigger than an original rownode is a dummy link
    selector = (row_idx > -1) & (row_idx < rows) & (col_idx > -1) & (col_idx < cols)
    row_idx = row_idx[selector]
    col_idx = col_idx[selector]

    return row_idx, col_idx
```

**Comment by fuglede** [comment]
Another related reduction that came up in https://github.com/networkx/networkx/discussions/4683 is that from the minimum weight matching where you make no assumptions at all on the size of the matching and you just want the best one overall. This can be achieved like the above by just using zeros as weights on _all_ new edges in the doubled graph.

**Comment by fuglede** [comment]
A possible API for those options could be

```
min_weight_bipartite_matching(...)
min_weight_maximum_bipartite_matching(...)
min_weight_full_bipartite_matching(...)
```
or as a `target_size` method parameter
```
min_weight_bipartite_matching(biadjacency_matrix, maximize=False, target_size='any')
```
with allowed values of `target_size` being `{'any', 'maximum', 'full'}` respectively. Main downside of the latter approach is that we would need a deprecation cycle to get rid of `min_weight_full_bipartite_matching`. An advantage would be being able to extend with `target_size` being an arbitrary integer. And avoiding the long names.

The only tricky part, as I see it, is handling the same issue we're talking about here: What to do with explicit zeros. For `'maximum'` and `'full'` we can push the problem to the user like we're doing now, but for `'any'`, i.e. the case where there is no constraint on the cardinality of the matching, we do need to do something as one needs to put a cost on adding an additional edge to the matching, so we can not simply offset the entire matrix. Moreover, 0 is the only sensible default cost. To me, this all suggests adding `min_weight_bipartite_matching` and taking care to handle explicit zeros even in the `target_size='full'` case, then deprecate `min_weight_full_bipartite_matching` if we feel its presence adds too much duplication.

**Comment by yalwan-sage** [comment]
Is there any technical reason why we can't always do the bound-bumping regardless of methodology? 

(referring to this:
```python
new_bigraph.data = new_bigraph.data - numpy.min(new_bigraph.data) + 1
```
)

If we're concerned that users have accidentally left in zeros they don't mean to you could do something like send a warning which can be disabled by a flag to the function which is basically "yes, i did mean to have zeros in here"? Or is there something else wider than this?

**Comment by fuglede** [comment]
No, you're right, the main complication is in ensuring we don't accidentally eliminate zeros, and that you've already taken care of. In fact, with all of that in place, and note that the underlying `_lapjvsp` is completely happy with explicit zeros, at the end of the day, it's not too much work to support explicit zeros at all.

If we were to introduce `min_weight_bipartite_matching(target_size='full')`, would it be very confusing if it supported explicit zeros when the existing one doesn't? Otherwise that might be a way to go. Then, if relevant, deprecate the old one (at the expense of users having to explicitly call `eliminate_zeros` if that's an actual use case).

We do have a [warning in place](https://github.com/scipy/scipy/blob/master/scipy/sparse/csgraph/_matching.pyx#L462-L463) already now (but it can't be toggled, outside of `filterwarnings` anyway).

**Comment by yalwan-sage** [comment]
> it's not too much work to support explicit zeros without artificially translating the entries at all.

Right but I think some of the docs mentioned issues handling zeros when converting between sparse representations. Is it worth keeping some caution around this? 

> would it be very confusing if it supported explicit zeros when the existing one doesn't?

I would say probably not, since it is an own method, provided the docs makes a very obvious effort to draw attention to this, something in bold under its own section or something. (a bit like me shouting not to call eliminate zeros in code comments 😅)
